File: backend/schemas/reports.py
from __future__ import annotations

from datetime import date
from typing import List

from pydantic import BaseModel

from core.enums import ReportPeriod


class ReportsSummary(BaseModel):
    income_minor: int
    expense_minor: int
    balance_minor: int
    savings_ratio_percent: float


class ExpenseByCategoryItem(BaseModel):
    category_id: str
    category_name: str
    amount_minor: int
    color_hex: str
    percentage: float


class IncomeExpenseSeriesItem(BaseModel):
    bucket_label: str
    income_minor: int
    expense_minor: int


class CashFlowSeriesItem(BaseModel):
    date: date
    balance_minor: int


class ReportsInsights(BaseModel):
    top_expense_category: str | None
    top_expense_amount_minor: int
    average_daily_expense_minor: int


class ReportsOverviewResponse(BaseModel):
    period: ReportPeriod
    start_date: date
    end_date: date
    summary: ReportsSummary
    expense_by_category: List[ExpenseByCategoryItem]
    income_expense_series: List[IncomeExpenseSeriesItem]
    cash_flow_series: List[CashFlowSeriesItem]
    insights: ReportsInsights

File: frontend/src/app/lib/api/reports.ts
import { apiRequest } from "./client";

export type ReportPeriod = "week" | "month" | "quarter";

export interface ReportsSummary {
  income_minor: number;
  expense_minor: number;
  balance_minor: number;
  savings_ratio_percent: number;
}

export interface ReportsExpenseByCategoryItem {
  category_id: string;
  category_name: string;
  amount_minor: number;
  color_hex: string;
  percentage: number;
}

export interface ReportsIncomeExpensePoint {
  bucket_label: string;
  income_minor: number;
  expense_minor: number;
}

export interface ReportsCashFlowPoint {
  date: string;
  balance_minor: number;
}

export interface ReportsInsights {
  top_expense_category: string | null;
  top_expense_amount_minor: number;
  average_daily_expense_minor: number;
}

export interface ReportsOverviewResponse {
  period: ReportPeriod;
  start_date: string;
  end_date: string;
  summary: ReportsSummary;
  expense_by_category: ReportsExpenseByCategoryItem[];
  income_expense_series: ReportsIncomeExpensePoint[];
  cash_flow_series: ReportsCashFlowPoint[];
  insights: ReportsInsights;
}

export function getReportsOverview(period: ReportPeriod, anchorDate: string) {
  return apiRequest<ReportsOverviewResponse>("/reports/overview", {
    method: "GET",
    params: {
      period,
      anchor_date: anchorDate,
    },
  });
}